gossip/src/weighted_shuffle.rs

```diff
@@ -411,6 +446,78 @@ mod tests {
         assert_eq!(shuffle.first(&mut rng), Some(17));
     }
 
+    // Verifies that changes to the code or dependencies (e.g. rand or
+    // rand_chacha::ChaChaRng) do not change the deterministic shuffle.
+    #[test_case(0x587c27258191c66d, "84jN8bvnp6mvtngzt42SW8AtRf5fcv3VBerKkUsYrCVG")]
+    #[test_case(0x7dad2afc68808779, "25oFhs9sR3WYfB6ohy752JrbLqpBjw6X4Eszbcsoxon4")]
+    #[test_case(0xfdd71c99c936736c, "7H9H8V7ccmpBhC3i5vEeFfiUwvRSAvRWadZhFH5ecSD7")]
+    #[test_case(0xe2a4d9fdd186636c, "Nxe6X7f74kEPrJFycKFcxByDRWKJtx1J3vsdbum9VPv")]
+    #[test_case(0x19a0a360e9f3094d, "Ec6wiaqDuVc5AzZpq4GAZ6GLsRJvw9mAVWVrCpDoGaRm")]
+    #[test_case(0xc5e0204894ca50dc, "BqxDzSFw8rJRHnTZmsPRzF77G3xgfK4hD8JyYeAFfxuZ")]
+    #[test_case(0xf1336cf933eeda07, "3Ux2vciDFdgNqULpsQpXfpaxZykWmBFCseqX9dwpGnyH")]
+    #[test_case(0xe666e7514f37c7a1, "Fc3gAUgh2mD1se3kkhPnLMKpQCiARd2PSdGf7b2fDS2n")]
+    fn test_weighted_shuffle_hard_coded_paranoid(seed: u64, expected_hash: &str) {
+        let expected_hash = Hash::from_str(expected_hash).unwrap();
+        let mut rng = <[u8; 32]>::try_from(
+            successors(Some(seed), |seed| Some(seed + 1))
+                .map(u64::to_le_bytes)
+                .take(32 / 8)
+                .flatten()
+                .collect::<Vec<u8>>(),
+        )
+        .map(ChaChaRng::from_seed)
+        .unwrap();
+        let num_weights = rng.gen_range(1..=100_000);
+        assert!((8143..=85348).contains(&num_weights), "{num_weights}");
+        let weights: Vec<u64> = repeat_with(|| {
+            if rng.gen_ratio(1, 100) {
+                0u64 // 1% zero weights.
+            } else {
+                rng.gen_range(0..=(u64::MAX / num_weights as u64))
+            }
+        })
+        .take(num_weights)
+        .collect();
+        let num_zeros = weights.iter().filter(|&&w| w == 0).count();
+        assert!((72..=846).contains(&num_zeros), "{num_zeros}");
```

Comment:
These are very particular numbers, where do they come from?

Reply:
Basically the resulting min and max num-zeros based on the hard coded seeds.
Just a minor sanity checks that the num-zeros looks reasonable.